UPS PUSH SDK

wtrocki opened the pull request:
## Motivation

Provide initial idea how Cordova Push SDK should look like.

Add base wrapper for Cordova plugin that will be used to stage changes that are not merged/released upstream. Add push registration method that can be used in cordova.

## Tasks

- [x] Registration plugin
- [x] Plugin wrapper

pb82 commented:
Do you also want to wrap the `registerDevice` event in there (allow users to provide a handler for that)? Or does this just expose the plugin itself? In that case, what's the purpose of the wrapper?

wtrocki commented:
@pb82 This wrapper will be used to lock the version of the plugin and being able to replace it in case of issues with something else. JS package will be used only for registration.

wtrocki commented on packages/push/src/index.ts:
```diff
@@ -0,0 +1,11 @@
+ *
+ * @param config object containing SDK. Object can be read from file or pasted directly in source code.
+ */
+export const init = (config: AeroGearConfiguration): void => {
```
Yeah. Going to push this changes soon.

wtrocki commented:
@pb82 This branch should work end to end with the plugin. 
I can push this changes to upstream for better collaboration if neeed.

danielpassos commented on packages/push-cordova/README.md:
```diff
@@ -0,0 +1,19 @@
+## AeroGear Metrics Cordova plugin
```
s/Metrics/Push

danielpassos commented on packages/push-cordova/README.md:
````diff
@@ -0,0 +1,19 @@
+## Installation
+
+```
+cordova plugin add @aerogear/cordova-plugin-aerogear-push
````
Any reason why not use the conversation aerogear-[plataform]-[feature]?

eg: aerogear-cordova-push-plugin

wtrocki replied:
If you mean `convention` we use cordova plugin convention for discoverability in npm.

danielpassos commented on packages/push-cordova/package.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "@aerogear/cordova-plugin-aerogear-push",
+  "version": "0.3.0",
```
0.3.0?

wtrocki replied:
We use single version per entire SDK. New packages will get the version that currently exist.

danielpassos replied:
It means all SDK should have the same version?

wtrocki replied:
We do not care as this version is being changed by our automation scripts on release. Then we decide what versions all packages should have.

danielpassos commented on packages/push/README.md:
```diff
@@ -0,0 +1,5 @@
+## AeroGear Services Unified Push Registration SDK
```
I think we should remove _Registration_ because it's not _only_  about _Registration_. Wdyt?

wtrocki replied:
It is only about registration, all other functionalities will be handled in the plugin. The whole purpose of our SDK should be to enable/simplyfy UPS registration

danielpassos commented on packages/push/package.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "@aerogear/push",
+  "version": "0.3.0",
```
0.3.0?

danielpassos commented on packages/push/src/PushRegistration.ts:
```diff
@@ -0,0 +1,82 @@
+      console.warn("Push configuration is missing. UPS server registration will not work.");
+      this.pushConfig = {};
+    } else {
+      this.pushConfig = configuration[0];
```
I really don't like this option, I see we also have this on other platforms, IMO the developer should provide the ID of the push he/she wanna use.

wtrocki replied:
@danielpassos This behavior was agreed on mailing list. Feel free to post this there. We cannot change this without changing other SDK's etc.

pb82 commented on packages/push/src/PushRegistration.ts:
```diff
@@ -0,0 +1,81 @@
+      return Promise.reject("Registration requires cordova plugin. Verify the " +
+        "@aerogear/cordova-plugin-aerogear-metrics plugin is installed.");
+    }
+    if (!this.pushConfig && !this.pushConfig.config && !this.pushConfig.url) {
```
Should this be `||` instead of `&&`? What if it does have `url` but not `config`? This condition would not fail in such a case.

pb82 commented on packages/push/src/PushRegistration.ts:
```diff
@@ -0,0 +1,81 @@
+        "@aerogear/cordova-plugin-aerogear-metrics plugin is installed.");
+    }
+    if (!this.pushConfig && !this.pushConfig.config && !this.pushConfig.url) {
+      Promise.reject("UPS registration: configuration is invalid");
```
missing `return`?

pb82 commented on packages/push/src/PushRegistration.ts:
```diff
@@ -0,0 +1,81 @@
+      return Promise.reject("UPS registration: Platform is configured." +
+        "Please add UPS variant and generate mobile - config.json again");
+    }
+    const authToken = window.btoa(`${platformConfig.variantId}:${platformConfig.variantSecret}`);
```
Are you sure that the whole string should be encoded and not only the individual parts? Looking at other UPS requests it could be that what you need is more like:

```
const auhtToken = `${window.btoa(platformConfig.variantId)}:${window.btoa(platformConfig.variantSecret)}`;
```

wtrocki replied:
I reverted that changes. So this comment appeared again. It works this way.

wtrocki commented:
@pb82 @darahayes This is ready for review. I have tested this changes on IOS platform.

wtrocki commented:
@pb82 I need to add unit tests and then going to merge it.

pb82 commented:
@wtrocki Ok, can we also publish this version to npm?

wtrocki commented:
@pb82 @darahayes Added tests. Added additional check if device token is empty. Fixed bug for promises (wrapped into error). IMHO this is really solid now. Can you do quick take look.

wtrocki commented:
@pb82 I think we can release it.
